### ISSUE TITLE
SSL connection fails when setup according to docs w/ self-signed cert (SG 1.5)

### ISSUE THREAD

**Issue by rajagp** [couchbase/sync_gateway#3126]
Attempted to configure Sync Gateway with SSL using a self signed certificate but connection fails. Followed instructions in https://developer.couchbase.com/documentation/mobile/1.5/guides/sync-gateway/configuring-ssl/index.html

Related Issue also reported https://forums.couchbase.com/t/todo-2-0-and-blips/15064/14
### Sync Gateway version
1.5.1

### Operating system
Mac OS . Issue also reported on Windows

### Config file
```javascript
"SSLCert": "ssl/cert.pem",
"SSLKey": "ssl/privkey.pem",
```

### Log output
```
2017-12-07T15:51:08.169-05:00 HTTP: Protocols enabled: [http/1.1] on 127.0.0.1:4985
2017-12-07T15:51:08.172-05:00 HTTP: Protocols enabled: [http/1.1] on :4984
```
 
### Expected behavior
Expect for connection to be established

### Actual behavior
Cannot connect to SGW on https

### Steps to reproduce
Tried the following (multi part message)
Case 1:

```
1. curl -X GET   https://localhost:4984/   -H 'cache-control: no-cache' --verbose
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4984 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /Users/priya.rajagopal/anaconda/ssl/cacert.pem
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
* Curl_http_done: called premature == 1
* Closing connection 0
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

Case 2:
```
curl -k  https://localhost/api/v1/status --verbose --sslv2

curl -k  https://localhost/api/v1/status --verbose --sslv2
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 443 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connection failed
* connect to 127.0.0.1 port 443 failed: Connection refused
* Failed to connect to localhost port 443: Connection refused
* Closing connection 0
curl: (7) Failed to connect to localhost port 443: Connection refused
```

Case3:
Tested via Safari browser as well.

**Comment by tleyden** [comment]
@raghusarangapani @sridevi-15 do any of the QE functional tests run with HTTPS/SSL?  If so, can you post a link so that we can check to see if there is any difference in the config compared to what's in the documentation?

**Comment by adamcfraser** [comment]
@rajagp Does `curl -k https://localhost:4984/` work?  I'm curious whether this is a CURL config problem, or a SG config problem.

**Comment by rajagp** [comment]
@adamcfraser I just used the "insecure" mode. Is there a different option I should use for self signed cert .
-k, --insecure
https://curl.haxx.se/docs/manpage.html
> (TLS) By default, every SSL connection curl makes is verified to be secure. This option allows curl to proceed and operate even for server connections otherwise considered insecure.
> 
> The server connection is verified by making sure the server's certificate contains the right name and verifies successfully using the cert store.

**Comment by adamcfraser** [comment]
Your examples above don't show you using -k for Case 1.  That works for me locally following the SG docs.  I haven't yet tried setting up CURL to use my self-signed cert - wanted to check first whether Case 1 worked for you w/ -k.

**Comment by raghusarangapani** [comment]
There are no tests that use SSL for SG. @sridevi-15 tried to set up SSL between server and SG and ran into a lot of issues.

**Comment by rajagp** [comment]
@adamcfraser This is what I see for Case1 with "-k" flag (Actually I had tried it earlier but copied the wrong example  for Case2 and didn't realize it when you commented earlier !)
```
curl -k -X GET   https://localhost:4984/   -H 'cache-control: no-cache' --verbose
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4984 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /Users/priya.rajagopal/anaconda/ssl/cacert.pem
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
* Curl_http_done: called premature == 1
* Closing connection 0
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

**Comment by adamcfraser** [comment]
I followed the steps in the docs, and got the following result with the same curl (after cleaning up some whitespace):
```
$ curl -k -X GET https://localhost:4984   -H 'cache-control: no-cache' --verbose
* Rebuilt URL to: https://localhost:4984  /
*   Trying ::1...
* Connected to localhost (::1) port 4984 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: localhost
> GET / HTTP/1.1
> Host: localhost:4984
> User-Agent: curl/7.43.0
> Accept: */*
> cache-control: no-cache
> 
< HTTP/1.1 200 OK
< Content-Length: 151
< Content-Type: application/json
< Server: Couchbase Sync Gateway/2.0 branch/feature/issue_3108 commit/a83c769+CHANGES
< Date: Fri, 08 Dec 2017 06:09:16 GMT
< 
* Connection #0 to host localhost left intact
{"couchdb":"Welcome","vendor":{"name":"Couchbase Sync Gateway","version":"2.0"},"version":"Couchbase Sync Gateway/feature/issue_3108(a83c769+CHANGES)"}
```

**Comment by rajagp** [comment]
Have it working my end as well. Comparing the curl output  from your end and mine, I noticed that the "server certificate" in my output  was not "localhost" (which was weird because I did have FQDN as localhost). So this was the hint that the certificate getting loaded was not OK.  

Double checking, I noticed that in addition to my manual sync gateway instance that I had launched ,  I had a docker instance running as well which was not SSL enabled and it appears that the curl request was hitting the docker one. So I stopped the docker instance and was able to connect to https just fine .
Of course still not sure why the user in the forum is not able to connect. I will try certificate pinning to see if the apps connect but that may be a separate issue . So I would think that SGW is good. Thanks @adamcfraser !
